sparse/coo/src/umath.py

```diff
@@ -0,0 +1,44 @@
+import numpy as np
+from numba import jit
+
+
+@jit(nopython=True)
+def pywhere(a, b):
+    """
+    Finds all indexes into a and b such that a[i] = b[j]. The outputs are sorted
+    in lexographical order.
+
+    Parameters
+    ----------
+    a, b : np.ndarray
+        The input 1-D arrays to match. If matching of multiple fields is
+        needed, use np.recarrays. These two arrays must be sorted.
+
+    Returns
+    -------
+    a_idx, b_idx : np.ndarray
+        The output indices of every possible pair of matching elements.
+    """
+    if len(a) == 0 or len(b) == 0:
+        return np.empty(0, dtype=np.uintp), np.empty(0, dtype=np.uintp)
+
+    a_ind, b_ind = [], []
+    nb = len(b)
+    ib = 0
+    match = 0
+
+    for ia, j in enumerate(a):
+        if j == b[match]:
+            ib = match
+
+        while ib < nb and j >= b[ib]:
+            if j == b[ib]:
+                a_ind.append(ia)
+                b_ind.append(ib)
+
+                if b[match] < b[ib]:
+                    match = ib
+
+            ib += 1
+
+    return np.array(a_ind, dtype=np.uintp), np.array(b_ind, dtype=np.uintp)
```

Comment:
Should we reduce the dtype here to be the minimum that can fit these values?

Reply:
There are a couple of reasons against that but I'm divided myself:

- We would have to give up nopython mode and go into mixed mode.
- There would be so many distinct signatures (all possible input dtypes, and different output dtypes) that the JIT-ing itself would take a bit of time.

For it are obviously:

- Better performance (maybe?) I haven't tested.
- Lower memory consumption.

I'll let you make the call.

Reply:
It seems I was right. The minimum dtype thing massively increases the time taken. 5.67 s for the benchmark in the comments. Writing a wrapper is still an option though.

Reply:
OK, happy to skip this for now